File: platforms/ios/DezelScanner/DezelScanner/View/ObjectExtractorView.swift
import AVFoundation
import Dezel

/**
 * @class ObjectExtractorView
 * @since 0.1.0
 */
open class ObjectExtractorView : UIView {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * The image that contains the soon to be extracted image.
	 * @property image
	 * @since 0.1.0
	 */
	open var image: UIImage? {
		didSet {
			self.background.image = image
		}
	}

	/**
	 * @property background
	 * @since 0.1.0
	 * @hidden
	 */
	private var background: UIImageView = UIImageView(frame: .zero)

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.1.0
	 */
	public override init(frame: CGRect) {
		super.init(frame: frame)
		self.background.contentMode = .scaleAspectFill
		self.addSubview(self.background)
	}

	/**
	 * @constructor
	 * @since 0.1.0
	 */
	required public init?(coder: NSCoder) {
		super.init(coder: coder)
		self.background.contentMode = .scaleAspectFill
		self.addSubview(self.background)
	}

	/**
	 * @inherited
	 * @method layoutSubviews
	 * @since 0.1.0
	 */
	open override func layoutSubviews() {
		self.background.frame = self.bounds
	}

	/**
	 * @method extract
	 * @since 0.1.0
	 */
	open func extract(p1: CGPoint, p2: CGPoint, p3: CGPoint, p4: CGPoint) -> UIImage? {

		guard let image = self.image?.cgImage else {
			return nil
		}

		let p1 = self.convert(p1, of: image)
		let p2 = self.convert(p2, of: image)
		let p3 = self.convert(p3, of: image)
		let p4 = self.convert(p4, of: image)

		if let extracted = DLExtractorPullImage(image, Int32(image.width), Int32(image.height), p1, p2, p3, p4) {
			return UIImage(pointer: extracted)
		}

		return nil
	}

	//--------------------------------------------------------------------------
	// MARK: Methods - Animations
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method action
	 * @since 0.1.0
	 */
	override open func action(for layer: CALayer, forKey event: String) -> CAAction? {
		return Transition.action(for: layer, key: event)
	}

	//--------------------------------------------------------------------------
	// MARK: Private API
	//--------------------------------------------------------------------------

	/**
	 * @method convert
 	 * @since 0.1.0
	 * @hidden
	 */
	private func convert(_ point: CGPoint, of image: CGImage) -> CGPoint {

		let backgroundW = self.background.bounds.width
		let backgroundH = self.background.bounds.height
		let imageW = CGFloat(image.width)
		let imageH = CGFloat(image.height)

		let ratioX = backgroundW / imageW
		let ratioY = backgroundH / imageH

 		let scale = max(ratioX, ratioY)

		var converted = point

        converted.x -= (backgroundW - imageW * scale) / 2.0
        converted.y -= (backgroundH - imageH * scale) / 2.0
     	converted.x /= scale
        converted.y /= scale

        return converted
	}
}

File: platforms/android/DezelScanner/lib/src/main/java/ca/logaritm/scanner/view/ObjectExtractorView.kt
package ca.logaritm.scanner.view

import android.content.Context
import android.graphics.Bitmap
import android.graphics.PointF
import android.view.ViewGroup
import android.widget.ImageView
import ca.logaritm.dezel.extension.Delegates
import ca.logaritm.scanner.scanner.ScannerExternal
import ca.logaritm.dezel.view.graphic.Convert

open class ObjectExtractorView(context: Context) : ViewGroup(context) {

	open var image: Bitmap? by Delegates.OnSetOptional<Bitmap>(null) { value ->
		this.background.setImageBitmap(value)
	}

	private var background: ImageView = ImageView(context)

	init {
		this.background.scaleType =  ImageView.ScaleType.CENTER_CROP
		this.addView(this.background)
	}

	/**
	 * @inherited
	 * @method onMeasure
	 * @since 0.1.0
	 */
	override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
		super.onMeasure(widthMeasureSpec, heightMeasureSpec)
		this.background.measure(widthMeasureSpec, heightMeasureSpec)
	}

	/**
	 * @inherited
	 * @method onLayout
	 * @since 0.1.0
	 */
	override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
		this.background.layout(l, t, r, b)
	}

	/**
	 * @method extract
	 * @since 0.1.0
	 */
	open fun extract(p1: PointF, p2: PointF, p3: PointF, p4: PointF): Bitmap? {

		val image = this.image
		if (image == null) {
			return null
		}

		val p1 = this.convert(p1, image)
		val p2 = this.convert(p2, image)
		val p3 = this.convert(p3, image)
		val p4 = this.convert(p4, image)

		return ScannerExternal.pullImage(image, image.width, image.height, p1, p2, p3, p4)
	}

	/**
	 * @method convert
	 * @since 0.1.0
	 */
	private fun convert(point: PointF, image: Bitmap): PointF {

		val backgroundW = Convert.toDp(this.background.width).toFloat()
		val backgroundH = Convert.toDp(this.background.height).toFloat()
		val imageW = image.width.toFloat()
		val imageH = image.height.toFloat()

		val ratioX = backgroundW / imageW
		val ratioY = backgroundH / imageH

		val scale = Math.max(ratioX, ratioY)

		var converted = point

		converted.x -= (backgroundW - imageW * scale) / 2f
		converted.y -= (backgroundH - imageH * scale) / 2f
		converted.x /= scale
		converted.y /= scale

		return converted
	}
}